kotlinx-coroutines-core/common/src/JobSupport.kt

```diff
@@ -517,8 +533,13 @@ public open class JobSupport constructor(active: Boolean) : Job, ChildJob, Paren
                 ?: InvokeOnCompletion(this, handler)
     }
 
-    private fun addLastAtomic(expect: Any, list: NodeList, node: JobNode<*>) =
-        list.addLastIf(node) { this.state === expect }
+    private fun addLast(expectedState: Any, list: NodeList, node: JobNode<*>): Boolean {
```

Comment:
This method not only adds to the tail, but does something else, please properly documented it and describe why this machinery is required. Especially `ChildCompletion`

Reply:
I've removed this extra logic from `addLast` method, it now just tries to add a node to the list.